package.json

```diff
@@ -39,28 +38,23 @@
   "homepage": "https://github.com/tinylibs/tinyexec#readme",
   "devDependencies": {
     "@eslint/js": "^9.0.0",
+    "@rollup/plugin-commonjs": "^26.0.1",
+    "@rollup/plugin-node-resolve": "^15.2.3",
+    "@rollup/plugin-typescript": "^11.1.6",
     "@types/cross-spawn": "^6.0.6",
     "@types/node": "^20.12.7",
     "c8": "^9.1.0",
     "cross-spawn": "^7.0.3",
     "eslint-config-google": "^0.14.0",
     "prettier": "^3.2.5",
-    "tsup": "^8.1.0",
+    "rollup": "^4.21.1",
+    "tslib": "^2.7.0",
     "typescript": "^5.4.5",
     "typescript-eslint": "^7.7.0"
   },
   "exports": {
-    ".": {
-      "import": {
-        "types": "./dist/main.d.ts",
-        "default": "./dist/main.js"
-      },
-      "require": {
-        "types": "./dist/main.d.cts",
```

Comment:
we should still end up with both of these (`main.d.ts` and `main.d.cts`) for typescript to be happy

what does rollup's typescript plugin output?

Reply:
I'm not exactly sure how to do that with rollup's typescript plugin, but I don't think it's necessary to have both since they have identical contents

Reply:
It is necessary to have both. If the package type is module, typescript treats all dts as esm types and vice versa. Requiring CTS for cjs modules

So they will be the same contents but hasn't to exist twice (TS is doing the right thing here, we're all wrong for making dual packages a thing 😅)

Reply:
hey guys I recently found out that ["Are The Types Wrong"](https://arethetypeswrong.github.io/?p=tinyexec%400.3.0) website have also created a CLI [`@arethetypeswrong/cli`](https://www.npmjs.com/package/@arethetypeswrong/cli) which allow you to test and run locally even before publishing (there's actually also the upload button on the website to upload a packed file that is also possible). 

@43081j are you sure that just cloning a TypeScript `d.ts` file to `.d.cts` with same content is enough? I think that I tried that before and the CLI above didn't pass, but maybe I've done something wrong!? Side note, I tried with plain tsc, not sure if the Rollup/esbuild DTS plugins bundles dts differently though, it looks like they do compared to tsc

Reply:
yes, attw will also test for this

basically you should have two exports in your `package.json`: one for `import` and one for `require`. each has its own entrypoint and its own `types` if the file can't be inferred

you can see an example in tinyexec:
https://unpkg.com/browse/tinyexec@0.3.0/package.json

attw is happy with this, whereas if we didn't have the cts and reused the same dts, it would complain about masquerading (correctly)

Reply:
but you're using tsup and yes I got it working with it, but if I use plain tsc and just copy/rename the d.ts to d.cts and update the import/require as you shown in tinyexec then it doesn't work. What I haven't tried though is to move esm/cjs into their own folders, that might avoid the masquerade detection of attw. Thanks for the info anyway

Reply:
tsup or not, you need two copies of the dts file (preferably a dts and a cts if the overall package is an es module) 

I'm not sure why it didn't work for you. You can run tsc twice or copy the files and achieve the same result. I've done that in other packages without tsup

Overall point being you can't use one dts file for a dual package, as that would be interpreted as the root type (e.g. esm if `type` is `module`)

Reply:
I figured out my problem, I wasn't using `.js` extensions in my imports and when testing my hybrid approach with `type=module` and passing it to attw it was saying that my files were the wrong format. So anyway that's fixed by using extensions and for the `d.cts` file, I simply added a simple Node script via `clone-dts.mjs`

```ts
// clone-dts.mjs
import { readFileSync, writeFileSync } from 'node:fs';

writeFileSync('dist/index.d.cts', readFileSync('dist/index.d.ts'));
```
isn't this the only thing missing for this PR (reverting back to the original `exports`)? There's probably a way to do it with a Rollup plugin, but we probably don't need to go crazy about it, my script works great with `&& node clone-dts.mjs` in my `build` script. I guess this Rollup plugin [rollup-plugin-copy](https://github.com/vladshcherbin/rollup-plugin-copy) would also work

Reply:
you would have to copy both the source and the DTS file

js/DTS and cjs/cts (2x everything)

I've lost track of what we're discussing here too since we've gone on a tutorial-like tangent 😅 

basically we need to make attw happy and we do that by having 2x everything. how you achieve that, I don't think matters much, whatever works. I don't think Ben has looked at this for a while anyway now, I'd probably have to pick it up at some point

Reply:
ah sorry, I just said that I got it all working on my end and I was actually discussing this PR in my last comment. I should have been more clearer 😆